### PR TITLE
Intel GPU support initialization

### PR DESCRIPTION
**Pull request by abhilash1910**
Intel GPU Support for Accelerate Framework .
Initial Draft PR :
-Feature addition for XPU devices backend
- CCL bindings for XPU devices
- Sanity checks
PR is in progress

### CODE REVIEW

File: src/accelerate/test_utils/scripts/test_script.py
Comment (sgugger) [inline src/accelerate/test_utils/scripts/test_script.py:57]: To clean up

File: src/accelerate/test_utils/scripts/test_sync.py
Comment (sgugger) [inline src/accelerate/test_utils/scripts/test_sync.py:26]: Should be removed.

File: src/accelerate/accelerator.py
Comment (sgugger) [inline src/accelerate/accelerator.py:385]:
```suggestion
                if is_xpu_available():
                    self.scaler = torch.xpu.amp.GradScaler(**kwargs)
                else:
                    self.scaler = torch.cuda.amp.GradScaler(**kwargs)
```

File: src/accelerate/accelerator.py
Comment (sgugger) [inline src/accelerate/accelerator.py:2458]: Same logic here, don't set `autocast_context` to then change it.

File: src/accelerate/accelerator.py
Comment (sgugger) [inline src/accelerate/accelerator.py:2462]: Same here.

File: src/accelerate/accelerator.py
Comment (sgugger) [inline src/accelerate/accelerator.py:2466]: Same there/

File: src/accelerate/hooks.py
Comment (sgugger) [inline src/accelerate/hooks.py:536]: Should probably be before the CUDA line.

File: src/accelerate/state.py
Comment (sgugger) [inline src/accelerate/state.py:114]:
```suggestion
                    if is_xpu_available():
                        torch.xpu.set_device(self.device)
                    else:
                        torch.cuda.set_device(self.device)
```

File: src/accelerate/state.py
Comment (sgugger) [inline src/accelerate/state.py:111]:
```suggestion
                        if is_xpu_available():
                            self.device = torch.device("xpu", self.local_process_index)
                        else:
                            self.device = torch.device("cuda", self.local_process_index)
```

File: src/accelerate/state.py
Comment (sgugger) [inline src/accelerate/state.py:147]: See comments above.

File: src/accelerate/test_utils/scripts/test_cli.py
Comment (sgugger) [inline src/accelerate/test_utils/scripts/test_cli.py:2]: This import is not used, and we can't make `accelerate test` have a hard dependency on ipex, so this should be removed.

File: src/accelerate/test_utils/scripts/test_cli.py
Comment (sgugger) [inline src/accelerate/test_utils/scripts/test_cli.py:15]: Should be removed as well, probably.

File: src/accelerate/utils/megatron_lm.py
Comment (sgugger) [inline src/accelerate/utils/megatron_lm.py:308]: Same comment as before. Don't initialize a variable on GPU to then erase it on put on XPU. Also not sure Megatron LM will support torch xpu so I don't think this is necesary.

File: src/accelerate/utils/memory.py
Comment (sgugger) [inline src/accelerate/utils/memory.py:58]: Protect this by a `is_xpu_available` instead (here and the two other below).

File: src/accelerate/utils/modeling.py
Comment (sgugger) [inline src/accelerate/utils/modeling.py:330]: This will replace `max_memory` on any setup having GPUs but not XPUs, so we can't change the file like this.

**Review by sgugger — COMMENTED** [review]
Thnaks for opening this PR. I think it tries to do way too much at the same time and should be split in smaller blocks: we don't need to have support for DeepSpeed +XPUs (is that even a thing?) or MegatronLM + XPUs (same) or big model inference on XPUs all in the same PR as enabling XPUs for distributed training.

I suggest starting smaller by just pushing support for distributed training, then we can add the other features if there is interest from the community.

**Comment by abhilash1910** [comment]
Hi @sgugger , the PR is still in debug state, we are working internally to ensure it functionally works. I will make the recommended changes in the meantime , before I assign a reviewer.

File: src/accelerate/utils/constants.py
Comment (abhilash1910) [reply src/accelerate/utils/constants.py:31]: yes this should be cleaned

File: src/accelerate/test_utils/scripts/test_cli.py
Comment (abhilash1910) [reply src/accelerate/test_utils/scripts/test_cli.py:15]: The test scripts will be  cleaned. (includes test code changes inside test_utils)

File: src/accelerate/test_utils/scripts/test_distributed_data_loop.py
Comment (abhilash1910) [reply src/accelerate/test_utils/scripts/test_distributed_data_loop.py:27]: Yes this would be removed and cleaned.

File: src/accelerate/state.py
Comment (abhilash1910) [reply src/accelerate/state.py]: So something like :
self.device = torch.device("xpu",self.local_process_index) if is_xpu_available() else torch.device("cuda",self.local_process_index)

File: src/accelerate/accelerator.py
Comment (mingxiaoh) [reply src/accelerate/accelerator.py:385]: sorry but there is no this api "torch.xpu.amp.GradScaler"

**Comment by abhilash1910** [comment]
@sgugger do you suggest to remove Megatron LM support for XPU ? I guess that seems ok considering the PR first tries to enable distributed training

File: src/accelerate/utils/modeling.py
Comment (abhilash1910) [reply src/accelerate/utils/modeling.py:385]: Resolved for now.

File: src/accelerate/commands/env.py
Comment (mingxiaoh) [inline src/accelerate/commands/env.py:1]: remove "[" please,

File: src/accelerate/utils/launch.py
Comment (mingxiaoh) [inline src/accelerate/utils/launch.py:107]: IndentationError: unindent does not match any outer indentation level

File: src/accelerate/utils/imports.py
Comment (mingxiaoh) [inline src/accelerate/utils/imports.py:113]: should be torch.xpu.is_available()

File: src/accelerate/utils/imports.py
Comment (mingxiaoh) [inline src/accelerate/utils/imports.py:114]: there is no this api  torch.xpu.is_bf16_supported() at all

**Review by mingxiaoh — APPROVED** [review]

File: src/accelerate/hooks.py
Comment (mingxiaoh) [inline src/accelerate/hooks.py:25]: why we need to touch mps?

Comment (abhilash1910) [reply src/accelerate/hooks.py:25]: This was already there ?

**Comment by sgugger** [comment]
As I said before, please keep the changes minimal to progressively enable support in small increments that are easy to review. The first integration should focus only on a small training on XPU. It shouldn't touch things like big model inference, megatron LM etc.

**Comment by abhilash1910** [comment]
@sgugger @pacman100   Is there a plan to support comm backend of deepspeed for deepspeed invocation ("ACCELERATE_USE_DEEPSPEED") ? Or only through torch.distributed permanently? ( https://github.com/huggingface/accelerate/pull/1138 , https://github.com/microsoft/DeepSpeed/issues/2117)

**Comment by sgugger** [comment]
@abhilash1910 We can discuss this in another PR. I feel you are not listening to the feedback we are given you here. Let's first support simple trainings on XPUs **then** we can iterate to add support for more integrations.

**Comment by abhilash1910** [comment]
sure @sgugger  , we are finalizing the changes on our end to just support simple trainings on XPU. Since state.py is foundational , hence thought of asking about it.

**Comment by abhilash1910** [comment]
@sgugger requesting review . As per the directives, only simplified distributed training on XPUs is enabled on this PR; do let us know on this.

File: src/accelerate/accelerator.py
Comment (sgugger) [inline src/accelerate/accelerator.py:1650]: This is not something we can do magically behind the scenes for the user. You should raise an error her if this can't be used for evaluation.

File: src/accelerate/accelerator.py
Comment (sgugger) [inline src/accelerate/accelerator.py:1654]: We should re-use the normal mixed precision argument here.

File: src/accelerate/checkpointing.py
Comment (sgugger) [inline src/accelerate/checkpointing.py:108]: This comment should be removed if it's not true anymore.

File: src/accelerate/checkpointing.py
Comment (sgugger) [inline src/accelerate/checkpointing.py:194]: Same here.

File: src/accelerate/commands/config/default.py
Comment (sgugger) [inline src/accelerate/commands/config/default.py:66]:
```suggestion
        num_xpus = torch.xpu.device_count()
```
Since they are called XPU everywhere...

File: src/accelerate/commands/env.py
Comment (sgugger) [inline src/accelerate/commands/env.py:59]:
```suggestion
        "PyTorch XPU available": str(pt_xpu_available),
```
No need to repeat the version here.

File: src/accelerate/utils/dataclasses.py
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:1360]: The dtype should be the mixed_precision argument we already have so shouldn't be here. Do you expect to have other arguments coming here? If no, there is no need for a plugin and we can simplify the integration.

File: src/accelerate/utils/imports.py
Comment (sgugger) [inline src/accelerate/utils/imports.py:212]: Let's use importlib here to have the check with `check_device=False` be cheap to execute.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for all your work here. I have flagged in comments a couple of places where we have breaking changes that really need to be addressed and left other comments on the style.

Could you also run `make style` on your branch to make sure the quality jobs pass?

File: src/accelerate/utils/imports.py
Comment (abhilash1910) [reply src/accelerate/utils/imports.py:212]: Ok there is already a ```is_torch_version``` in versions.py which invokes the importlib which can be used.

File: src/accelerate/utils/dataclasses.py
Comment (abhilash1910) [reply src/accelerate/utils/dataclasses.py:1360]: FP16 mixed precision is not supported yet through xpu . Yes so for args if it comes in as FP16 it will raise an issue .

**Comment by abhilash1910** [comment]
@sgugger ran make style and addressed reviews. Could you take a look again ? (Have a doubt in one comment regarding model.train - https://github.com/huggingface/accelerate/pull/1118/#discussion_r1178174712) Thanks

File: src/accelerate/accelerator.py
Comment (sgugger) [reply src/accelerate/accelerator.py:1652]: Yes, so that the user has to manually do the `model.train()` and not be surprised.

**Review by sgugger — COMMENTED** [review]
Clarified my comment on the training, and I do insist on re-using the existing mixed_precision. You can check when the state is initialized and raise an error if there is a non-supported value for XPUs there (mixed precision cannot change once the state has been initialized).

File: src/accelerate/accelerator.py
Comment (abhilash1910) [reply src/accelerate/accelerator.py]: Got it , makes sense. I think  I can remove the model.train() aspect since it is based  on users' discretion. 
Update: Added commits

File: src/accelerate/accelerator.py
Comment (abhilash1910) [reply src/accelerate/accelerator.py]: Update: Yes should add normal Mix precision; added.

File: src/accelerate/utils/dataclasses.py
Comment (abhilash1910) [reply src/accelerate/utils/dataclasses.py]: Yes resolved in new commit. I agree the default mix precision should be used and flagged accordingly for device support.

**Comment by abhilash1910** [comment]
Thanks @sgugger  for the detailed review. updated the commits as per suggestions. (+ test npr /styling).
Could you take a look ?

**Review by sgugger — COMMENTED** [review]
Thanks a lot for addressing all of those comments every thing looks good apart from the XPUPlugin which was a question in my review from a couple of days ago: why have a plugin (both for XPU or IPEX) that only has one flag? And like for the MPS device, we want the XPUs to be used automatically if there are available (unless the user specified `use_cpu=True`) so that flag `use_xpu` should probably not be there as well to make the API easier to use (there initially was a `use_mps` flag we deprecated for context).

I think the whole `XPUPlugin` should just be removed, and if XPUs are detected, they are used (just like CUDA GPUs or MPS devices are automatically used when available). Unless there is a potential conflict between XPUs and CUDA GPUs? But it seems from what I read they are mutually exclusive.

File: src/accelerate/test_utils/scripts/check_main_process_first.txt
Comment (sgugger) [inline src/accelerate/test_utils/scripts/check_main_process_first.txt:1]: Is this used anywhere?

File: src/accelerate/state.py
Comment (sgugger) [inline src/accelerate/state.py:139]: This needs to be executed is `self.device` is not `None` and we have a CUDA device.
Likewise, the `torch.xpu.set_device(self.device)` line above probably needs to be executed if `self.device` is not `None` and we have an XPU device.

File: src/accelerate/state.py
Comment (sgugger) [inline src/accelerate/state.py:553]: This isn't passed anymore, so should be removed, no?

File: src/accelerate/state.py
Comment (sgugger) [inline src/accelerate/state.py:614]: Same here.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for iterating! We're very close to have this merged :-)
There are a couple of lines of code that are left with the XPU plugin in the state module that need to be cleaned up.

At this stage, I'd also like if @muellerzr could have a second look on the PR.

File: src/accelerate/test_utils/scripts/check_main_process_first.txt
Comment (muellerzr) [reply src/accelerate/test_utils/scripts/check_main_process_first.txt:1]: This is the extra txt file generated that https://github.com/huggingface/accelerate/pull/1381 got rid of. So should be removed fully before merging.

File: src/accelerate/accelerator.py
Comment (abhilash1910) [reply src/accelerate/accelerator.py]: resolved

File: src/accelerate/state.py
Comment (abhilash1910) [reply src/accelerate/state.py]: Resolved

File: src/accelerate/accelerator.py
Comment (abhilash1910) [reply src/accelerate/accelerator.py]: Added the change

**Comment by abhilash1910** [comment]
Thanks @sgugger  @muellerzr for pointing out the details. Let me know if this is correct?

**Review by muellerzr — APPROVED** [review]
This all looks good to me, however can we address the failing tests? :)

**Comment by abhilash1910** [comment]
@muellerzr  Thanks for reviewing . Fixed the error in checkpointing tests, but the other tests fail since ipex is not installed .
I also added a small change to allow xpu amp to be used if xpu devices are detected rather (prepare_ipex method). With ipex installed (with or without xpus) test passes.
Could you please review ?

File: src/accelerate/accelerator.py
Comment (sgugger) [inline src/accelerate/accelerator.py:1143]: Those checks now need to be adapted since the plugin is always there.

**Review by sgugger — COMMENTED** [review]
The tests shouldn't fail if `ipex` is not installed. The goal is not to add a hard dependency on `ipex` but to use it when it's available. On CPU, Ipex should be used if it's available only. On XPUs I think the error could be changed to a warning to install Ipex.

**Comment by abhilash1910** [comment]
@sgugger  @muellerzr seems like this commit should fix the tests associated (waiting for CI, but locally this CI test passes). 
Thanks again for the comments.

**Comment by jgong5** [comment]
> The tests shouldn't fail if `ipex` is not installed. The goal is not to add a hard dependency on `ipex` but to use it when it's available. On CPU, Ipex should be used if it's available only. On XPUs I think the error could be changed to a warning to install Ipex.

@sgugger XPU is still out of PyTorch tree and supported via IPEX. Before we upstream XPU support to PyTorch, its functionality depends on the availability of IPEX. Can we skip XPU tests if IPEX is not installed?

**Comment by abhilash1910** [comment]
@sgugger @muellerzr  seems like tests pass now ( no longer a hard dependency on ipex installation). 
Let us know on this? :)

**Comment by sgugger** [comment]
Yes, let us just run all slow tests on this and if all goes well it can be merged. Thanks a lot for all your efforts on this!

**Comment by muellerzr** [comment]
@abhilash1910 we have a number of tests failing:

On single GPU:
```
 FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_get_balanced_memory - AssertionError: {0: 215, 1: 300} != {0: 300, 1: 300}
- {0: 215, 1: 300}
?     ^^^

+ {0: 300, 1: 300}
?     ^^^
```

On multi-GPU:
```
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_bnb - AssertionError: False is not true
```
And most concerning, when testing the example scripts, it looks like the Accelerator is prioritizing the CPU over the GPU, even when ipex isn't available. This was tested on two T4's running the following:

```bash
CUDA_VISIBLE_DEVICES="0" pytest -sv tests/test_examples.py -k test_checkpointing_by_epoch
```
And in this script (located at `examples/by_feature/checkpointing.py` I changed lines 124-130 to be:
```diff
def training_function(config, args):
    # For testing only
    if os.environ.get("TESTING_MOCKED_DATALOADERS", None) == "1":
        config["num_epochs"] = 2
    # Initialize accelerator
    accelerator = Accelerator(cpu=args.cpu, mixed_precision=args.mixed_precision)
+    assert accelerator.device.type == "cuda", f'Device: {accelerator.device}, type: {accelerator.device.type}'
```

**Comment by abhilash1910** [comment]
@muellerzr is the failure only on tests with modelling.py ? Prioritisation is strange ,maybe due to checking for xpu before cuda in some places . I will see why this fails .

**Comment by muellerzr** [comment]
@abhilash1910 the two test files are there in the trace I gave, `test_modeling_utils` and `test_big_modeling`. All the example tests are failing silently because what should take 6 minutes to run to completion takes 30+ minutes

File: src/accelerate/utils/modeling.py
Comment (mingxiaoh) [inline src/accelerate/utils/modeling.py:491]: should change to if not (torch.cuda.is_available() or is_xpu_available()):

**Review by mingxiaoh — CHANGES_REQUESTED** [review]

**Comment by abhilash1910** [comment]
@muellerzr  does this still fail in the slow tests?

**Comment by muellerzr** [comment]
@abhilash1910 we're now failing our CPU tests (notice that `test_checkpoint_step` has been running for 2+ hours and not passing. Can you try running `make test_checkpoint_step` to see if it passes? I'll run the slow tests after we get an all green again. (This should take *no more* than 2-3 minutes to run)

**Comment by muellerzr** [comment]
@abhilash1910 still not using CUDA on the example tests. Did you try making my modification to the checkpointing script to make sure it was using CDUA in practice? (If you don't have a GPU to use, you could try using Google Colab to test)

**Comment by abhilash1910** [comment]
Hi @muellerzr  , I did some testing on colab for checkpointing.py (using the assertion for cuda which you gave) with the xpu wheel build(1.19.9 -dev build) , and when I specify args.cpu=None it does pick up the GPU cuda device . I am sharing a snapshot: 
image
The colab is here: https://colab.research.google.com/drive/15NZd13igA_S-jresSiKVcb6p2qXatziI?usp=sharing
This is also seen with the default version 1.19 of accelerate . So with cpu=None , it does pick up the cuda and assertion passes. Could you please check if anything is missing ? I will re-check again .

**Comment by muellerzr** [comment]
@abhilash1910 this also needs to work *without* the XPU build, which is where I believe we're facing the breaking issue :) As it still shows `cpu` for me

**Comment by abhilash1910** [comment]
@muellerzr  I had tried with the same notebook with the default public accelerate(1.19.0)  but saw the same results  on colab(standard public - 1.19.0 ; xpu wheel - 1.19.0-dev 0). (as mentioned here: https://github.com/huggingface/accelerate/pull/1118#issuecomment-1539592556) . Both the public and the xpu build picks up the cuda properly.  I put args.cpu=None so as to use GPU cuda device in both cases.(since colab has no xpu ). Could you suggest which scripts to check further? Thanks

**Comment by abhilash1910** [comment]
Thanks @muellerzr  for sharing the test notebook. It was a tricky issue  to resolve, and now it passes the slow tests (both checkpoint & and modeling utils) on T4 with cuda getting used.  Let me know on this ?

**Comment by muellerzr** [comment]
Great, slow tests are running

**Review by muellerzr — APPROVED** [review]
Tests are passing fine here! cc @sgugger for a final look

**Comment by muellerzr** [comment]
@abhilash1910 it seems this has broken something in diffusers, please see this issue: https://github.com/huggingface/accelerate/issues/1420

Any insight towards what could be going wrong here? This is breaking some of our other libraries

File: src/accelerate/accelerator.py
Comment (pacman100) [inline src/accelerate/accelerator.py:325]: Hello, removing this is leading to various test failures as we are trying to integrate accelerate in transformers trainer. We are planning to revert this line and wanted to check if there was a reason for removing this.

**Pull request merged**
